tools/src/launcher/Main.hs

```diff
@@ -178,16 +183,92 @@ reportErrorDefault filename contents = do
+    -- TODO: figure out a less ugly OS conditionalisation method
+    -- TODO: decide if we want escaping
+#ifdef mingw32_HOST_OS
+    pEnvvarRef   = P.between (P.char '%') (P.char '%') pEnvvarName <&> EnvVar . toText
+    pPassiveText = P.some (P.noneOf "%")                           <&> Plain  . toText
+#else
+    pEnvvarRef   = P.char '$' >>
+                   P.choice [ pEnvvarName
+                            , P.between (P.char '{') (P.char '}') pEnvvarName
+                            ] <&> EnvVar
+    pPassiveText = P.some (P.noneOf "$") <&> Plain  . toText
+#endif
+    pEnvvarName :: (Monad p, P.TokenParsing p) => p Text
+    pEnvvarName = (P.some $ P.choice [P.alphaNum, P.char '_']) <&> toText
+
+-- | Substitute environment variable references. The 'desc' argument supplies
+-- human-oriented context description in case of error.
+substituteEnvVarsText :: Text -> Text -> IO Text
+substituteEnvVarsText desc text = do
+  chunks <- case parseEnvrefs text of
+    P.Success r  -> pure r
+    P.Failure ei -> do
+      let err = desc <> "\n" <> (show $ P._errDoc ei)
+      reportErrorDefault "config-parse-error.log" err
+      error err
+  substd <- forM chunks $
+    \case Plain  x -> pure x
+          EnvVar x -> do
+            val <- lookupEnv $ toString x
+            case val of
+              Just x' -> pure $ toText x'
+              Nothing -> do
+                let err = desc <> "\n" <> "Reference to an undefined environment variable '"<> x <>"'"
+                reportErrorDefault "config-parse-error.log" err
+                error err
+  pure $ T.concat substd
+
+-- | Given an Aeson 'Value', parse and substitute environment variables in all
+--   'AE.String' objects.  The 'desc' argument supplies context in case of error.
+substituteEnvVars :: Text -> Value -> IO Value
+substituteEnvVars desc (AE.String text) = AE.String <$> substituteEnvVarsText desc text
+substituteEnvVars desc (AE.Array xs)    = AE.Array  <$> traverse (substituteEnvVars desc) xs
+substituteEnvVars desc (AE.Object o)    = AE.Object <$> traverse (substituteEnvVars desc) o
+substituteEnvVars _    x                = pure x
+
 getLauncherOptions :: IO LauncherOptions
 getLauncherOptions = do
     LauncherArgs {..} <- either parseErrorHandler pure =<< execParserEither programInfo
+#ifdef mingw32_HOST_OS
+    daedalusDir <- takeDirectory <$> getExecutablePath
+    -- This is used by 'substituteEnvVars', later
+    setEnv "DAEDALUS_DIR" daedalusDir
+#endif
     configPath <- maybe defaultConfigPath pure maybeConfigPath
     decoded <- Y.decodeFileEither configPath
     case decoded of
         Left err -> do
             reportErrorDefault "config-parse-error.log" $ show err
             throwM $ ConfigParseError configPath err
-        Right op -> expandVars op
+        Right value -> do
+          let contextDesc = "Substituting environment vars in '"<>toText configPath<>"'"
+          substituted <- substituteEnvVars contextDesc value
```

Comment:
We have `value` which comes from a yaml parse, and now we do substituion. But yaml parsing already does substitution! `YamlParser a ~ AnchorMap -> Either Text a`, where `AnchorMap ~ Map String YamlValue`. Perhaps we could finesse all of this and make use of the yaml parser's implementation? If we replace the environment-variable-style `$<name>` or `%<name>%` forms with yaml-style `*<name>` forms in the yaml text itself, and we runt he parser with a particular initial `AnchorMap` that includes the relevant environment variable names, then we'll get this all for free.

Reply:
```Haskell
envAnchorMap :: IO AnchorMap
envAnchorMap = fmap (M.fromList . second makeYamlValue) getEnvironment
  where
    makeYamlValue :: String -> YamlValue
    makeYamlValue str = Scalar (B8.pack str) StrTag SingleQuoted Nothing

-- Defined in Data.Yaml.Parser
data RawDoc = RawDoc YamlValue AnchorMap
    deriving Show

-- Defined in Data.Yaml.Parser
parseRawDoc :: (FromYaml a, MonadThrow m) => RawDoc -> m a
parseRawDoc (RawDoc val am) =
    case unYamlParser (fromYaml val) am of
        Left t -> throwM $ FromYamlException t
        Right x -> return x

-- Defined in Data.Yaml.Parser
readYamlFile :: FromYaml a => FilePath -> IO a
readYamlFile fp = runConduitRes (decodeFile fp .| sinkRawDoc) >>= parseRawDoc

-- Our custom variant:
readYamlRawDoc :: FromYaml a => FilePath -> IO a
readYamlRawDoc fp = do
    am <- envAnchorMap
    RawDoc yaml am' <- runConduitRes (decodeFile fp .| sinkRawDoc)
    -- Check union order semantics. Do we want env vars to overwrite anchors in the raw doc?
    prawRawDoc (RawDoc yaml (M.union am' am))
```

Reply:
Do we fully control the yaml? If so then sure we don't need to support $ and % style substitution at all and could use the built-in substitution.

Reply:
Thank you!